### PR TITLE
feat: implements pool price oracle

### PR DESCRIPTION
**Pull request by jinoosss**
## Descriptions

This PR implements a Time-Weighted Average Price (TWAP) oracle system for GnoSwap pools, similar to Uniswap V3's price oracle mechanism. The implementation provides accurate price feeds for any time period within the pool's history, enabling reliable price data for DeFi applications, trading strategies, and external integrations.

### **1. TWAP Price Oracle**
- **Time-Weighted Average Price**: Calculates average prices over specified time periods
- **High Precision**: Uses X96 precision for accurate price calculations
- **Historical Data**: Supports price queries for past time periods
- **Real-time Updates**: Automatically updates cumulative values on price changes

### **2. Cumulative Price Tracking**
- **Price Cumulatives**: Tracks cumulative sqrt price with time weighting
- **Liquidity Cumulatives**: Tracks cumulative liquidity for advanced calculations
- **Automatic Updates**: Updates on every swap operation
- **Timestamp Tracking**: Maintains precise timing for accurate calculations

### **3. Public API Functions**
- **`GetPriceObserve()`**: Get TWAP price for a specific time period
- **`GetSlot0PriceOracle()`**: Get current oracle state for validation
- **Error Handling**: Robust error handling for invalid inputs

#### **Pool TWAP Fields (`pool_type.gno`)**
```go
// TWAP related fields
lastCumulativeUpdateTime int64      // last cumulative update time (in seconds)
sqrtPriceCumulativeX96   *u256.Uint // cumulative sqrt price (Q64.96 precision)
liquidityCumulative      *u256.Uint // cumulative liquidity (time-weighted average calculation)
```

#### **1. Price Cumulative Calculation (`calculatePriceCumulatives()`)**
- Calculates time-weighted cumulative values
- Updates both price and liquidity cumulatives
- Handles zero time delta cases efficiently

#### **2. TWAP Observation (`observe()`)**
- Calculates TWAP price for specified time period
- Uses linear interpolation for past price estimation
- Validates time period constraints
- Returns X96 precision sqrt price

#### **3. Oracle State Management (`slot0PriceOracle()`)**
- Returns current oracle state for validation
- Updates cumulatives before returning state
- Provides debugging and verification capabilities

#### **Swap Integration (`swap.gno`)**
```go
// update TWAP state
pool.updatePriceCumulatives()
```
- Automatically updates TWAP state on every swap
- Ensures accurate price tracking
- No performance impact on swap operations

#### **Pool Initialization (`type.gno`)**
```go
lastCumulativeUpdateTime: time.Now().Unix(),
sqrtPriceCumulativeX96:   u256.Zero(),
liquidityCumulative:      u256.Zero(),
```
- Initializes TWAP fields on pool creation
- Sets proper initial values for new pools

#### **`GetPriceObserve(poolPath string, secondsAgo int64) string`**
- **Purpose**: Get TWAP price for a specific time period
- **Parameters**: 
  - `poolPath`: Pool identifier
  - `secondsAgo`: Time period in seconds
- **Returns**: TWAP price as string (X96 precision)
- **Error Handling**: Returns "0" on error

#### **`GetSlot0PriceOracle(poolPath string) (int64, string, string)`**
- **Purpose**: Get current oracle state for validation
- **Returns**: 
  - `lastCumulativeUpdateTime`: Last update timestamp
  - `sqrtPriceCumulativeX96`: Current cumulative sqrt price
  - `liquidityCumulative`: Current cumulative liquidity

### **TWAP Calculation Algorithm**

1. **Time Delta Calculation**: `currentTime - lastUpdateTime`
2. **Cumulative Update**: `cumulative += currentPrice * timeDelta`
3. **Past Estimation**: `pastCumulative = currentCumulative - (currentPrice * secondsAgo)`
4. **Price Delta**: `priceDelta = currentCumulative - pastCumulative`
5. **TWAP Calculation**: `twapPrice = priceDelta / secondsAgo`

### **Precision Handling**
- **X96 Precision**: All price calculations use Q64.96 fixed-point arithmetic
- **U256 Arithmetic**: Uses uint256 for overflow protection
- **Time Precision**: Unix timestamps for accurate time tracking

### **Error Handling**
- **Invalid Time Period**: Returns error for `secondsAgo <= 0`
- **Historical Limit**: Returns error if `secondsAgo > lastUpdateTime`
- **Zero Division**: Handles edge cases with zero time deltas
- **No History**: Returns current price when no historical data available

### CODE REVIEW

**Comment by dongwon8247** [comment]
can you add tests to ensure it's calculating correctly?

File: contract/r/gnoswap/pool/swap.gno [hunk 1]
```diff
@@ -106,6 +107,10 @@ func Swap(
 
 	applySwapResult(pool, result)
 
+	// update TWAP state
+	currentTime := time.Now().Unix()
+	pool.updatePriceCumulatives(currentTime)
```
Comment (notJoon) [inline contract/r/gnoswap/pool/swap.gno:112]: Don't we need to call it from `Mint` or `Burn`?

File: contract/r/gnoswap/pool/pool_type.gno [hunk 2]
```diff
@@ -121,3 +126,93 @@ func (p *Pool) Ticks() string {
...
+	// For simplicity, we use linear interpolation: past_cumulative = current_cumulative - (current_price * secondsAgo)
+	pastTime := lastCumulativeUpdateTime - secondsAgo
+	if pastTime < 0 {
+		pastTime = 0
+	}
+
+	// Estimate past cumulative by subtracting the price contribution over the time period
+	// past_cumulative = current_cumulative - (current_price * secondsAgo)
+	priceContribution := u256.Zero().Mul(p.slot0.sqrtPriceX96, u256.NewUintFromInt64(secondsAgo))
+	pastSqrtPriceCumulativeX96 := u256.Zero().Sub(currentSqrtPriceCumulativeX96, priceContribution)
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply contract/r/gnoswap/pool/pool_type.gno:196]: I think it's inefficient to accumulate history for observe,
remove this feature of observe itself and simplify it to only store price accumulation.

File: contract/r/gnoswap/pool/getter.gno [hunk 3]
```diff
@@ -126,3 +128,21 @@ func PoolGetTickSecondsOutside(poolPath string, tick int32) uint32 {
...
 	return mustGetPool(poolPath).GetTickInitialized(tick)
 }
+
+// TWAP related getter functions
+func GetPriceObserve(poolPath string, secondsAgo int64) string {
+	pool := mustGetPool(poolPath)
+	currentTime := time.Now().Unix()
+	twapPrice, err := pool.observe(secondsAgo, currentTime)
+	if err != nil {
+		return "0"
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply contract/r/gnoswap/pool/getter.gno:138]: Initialized when the pool is created
https://github.com/gnoswap-labs/gnoswap/pull/784/files/1710de23c18ffd5657cd4829e9f48cda36cd21e5#diff-f2826084d3cfe7d7dfed0350bb635695a07a0294bfb906a472f6aa43b87005f7R290-R292

File: contract/r/gnoswap/pool/swap.gno (hunk 1 above)
Comment (jinoosss) [reply contract/r/gnoswap/pool/swap.gno:112]: It doesn't matter, we only need to track the price change of the pool.

**Comment by dongwon8247** [comment]
1. Change to tickCumulative from priceCumulative (https://github.com/Uniswap/v3-core/blob/main/contracts/libraries/Oracle.sol)
2. Add event

File: contract/r/gnoswap/pool/pool_type.gno [hunk 4]
```diff
@@ -77,6 +77,11 @@ type Pool struct {
 	ticks                *avl.Tree  // tick(int32) -> TickInfo
 	tickBitmaps          *avl.Tree  // tick(wordPos)(int16) -> bitMap(tickWord ^ mask)(*u256.Uint)
 	positions            *avl.Tree  // maps the key (caller, lower tick, upper tick) to a unique position
+
+	// TWAP related fields
+	lastCumulativeUpdateTime int64      // last cumulative update time (in seconds)
+	sqrtPriceCumulativeX96   *u256.Uint // cumulative sqrt price (Q64.96 precision)
+	liquidityCumulative      *u256.Uint // cumulative liquidity (time-weighted average calculation)
```
[the remark this answers is not in the document]
Comment (jinoosss) [reply contract/r/gnoswap/pool/pool_type.gno:84]: Looks fine 👍

[the remark this answers is not in the document]
Comment (notJoon) [reply contract/r/gnoswap/pool/pool_type.gno]: https://github.com/gnoswap-labs/gnoswap/pull/784/commits/b52c25e52b79e0c24819334de0957bf9cd3a6452

**Comment by dongwon8247** [comment]
To be finalized by @jinoosss 🙏 

We still need to change to: 
```
    struct Observation {
        // the block timestamp of the observation
        uint32 blockTimestamp;
        // the tick accumulator, i.e. tick * time elapsed since the pool was first initialized
        int56 tickCumulative;
        // the seconds per liquidity, i.e. seconds elapsed / max(1, liquidity) since the pool was first initialized
        uint160 secondsPerLiquidityCumulativeX128;
        // whether or not the observation is initialized
        bool initialized;
    }
```

And add events that the backend needs to calculate and store TWAP

**Review by dongwon8247 — APPROVED** [review]
